ui/shared/url-util.js

```diff
@@ -0,0 +1,177 @@
+ * @class UrlUtil
+ */
+const UrlUtil = {
+
+  /**
+   * Extracts the scheme from a value.
+   * @param {String} input The input value.
+   * @returns {String} The found scheme.
+   */
+  getScheme(input) {
+    // This function returns one of following:
+    // - scheme + ':' (ex. http:)
+    // - scheme + '://' (ex. http://)
+    // - null
+    const scheme = (rscheme.exec(input) || [null])[0];
+    return scheme === 'localhost://' ? null : scheme;
+  },
+
+  /**
+   * Checks if an input has a scheme (e.g., http:// or ftp://).
+   * @param {String} input The input value.
+   * @returns {Boolean} Whether or not the input has a scheme.
+   */
+  hasScheme(input) {
+    return !!this.getScheme(input);
+  },
+
+  /**
+   * Prepends file scheme for file paths, otherwise the default scheme
+   * @param {String} input path, with opetional schema
+   * @returns {String} path with a scheme
+   */
+  prependScheme(input) {
+    // expand relative path
+    if (input.startsWith('~/')) {
+      input = input.replace(/^~/, os.homedir());
+    }
+
+    // detect absolute file paths
+    if (input.startsWith('/')) {
+      input = fileScheme + input;
+    }
+
+    // If there's no scheme, prepend the default scheme
+    if (!this.hasScheme(input)) {
+      input = defaultScheme + input;
+    }
+
+    return input;
+  },
+
+  /**
+   * Checks if a string is not a URL.
+   * @param {String} input The input value.
+   * @returns {Boolean} Returns true if this is not a valid URL.
+   */
+  isNotURL(input) {
+    // for cases, ?abc and "a? b" which should searching query
+    const case1Reg = /^(\?)|(\?.+\s)/;
+    // for cases, pure string
+    const case2Reg = /[\?\.\s\:]/;
+    // for cases, data:uri and view-source:uri
+    const case3Reg = /^\w+\:.*/;
+
+    let str = input.trim();
+    if (case1Reg.test(str) || !case2Reg.test(str) ||
+        this.getScheme(str) === str) {
+      return true;
+    }
+    if (case3Reg.test(str)) {
+      return false;
+    }
+
+    str = this.prependScheme(str);
+    const parsed = urlParse(str);
+    return !parsed.protocol || !parsed.host || !parsed.host.includes('.');
+  },
+
+  /**
+   * Converts an input string into a URL.
+   * @param {String} input The input value.
+   * @returns {String} The formatted URL.
+   */
+  getUrlFromInput(input) {
+    input = input.trim();
+
+    input = this.prependScheme(input);
+
+    if (this.isNotURL(input)) {
+      return input;
+    }
+
+    return urlParse(input).href;
+  },
+
+  /**
+   * Checks if a given input is a valid URL.
+   * @param {String} input The input URL.
+   * @returns {Boolean} Whether or not this is a valid URL.
+   */
+  isURL(input) {
+    return !this.isNotURL(input);
+  },
+
+  /**
+   * Checks if a URL is a view-source URL.
+   * @param {String} input The input URL.
+   * @returns {Boolean} Whether or not this is a view-source URL.
+   */
+  isViewSourceUrl(url) {
+    return url.toLowerCase().startsWith('view-source:');
```

Comment:
We should do `(url || "")` or default arg or something and not throw if we pass in undefined/null